### PR TITLE
Mito ai chat

### PR DESCRIPTION
**Pull request by naterush**
# Description

Implements: https://www.notion.so/trymito/AI-Chat-d2a69ef6b0344882851480efadd36bc9?pvs=4

It works really, really well. The main UX jumps are:
1. auto execute code
2. stay focused on the input

It leads to an _insanely_ fast way of getting things done in Mito. I honestly think this feature is 10x better, and I feel like I am going to use it a ton.

To do this, I added a new hook called `useSendEditOnClickNoParams`. It's documented, but TLDR is that we don't need to store params in the way we did previously with the useSendEditOnClick param - and we also want all the historical values of the params. 

Making a new hook was the _right_ call, as this results in much simpler code in the hook, rather than further complicating it.
 
# Testing

Use it. 

# Documentation

We need to redo these docs.

### CODE REVIEW

File: mitosheet/src/hooks/useEditOnClickNoParams.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        void refreshOnRedo();
+    }, analysisData)
+
+    // NOTE: all edit events are the name of the step + _edit
+    const editEvent = stepType + '_edit';
+
+    // This function actually sends the edit message to the backend
+    const edit = async (params: ParamType) => {
+
+        let newStepID = getRandomId();
```
Comment (naterush) [inline mitosheet/src/hooks/useEditOnClickNoParams.tsx:48]: set to const

File: mitosheet/src/components/taskpanes/AITransformation/AITransformationTaskpane.tsx [hunk 2]
```diff
@@ -255,161 +174,128 @@ const AITransformationTaskpane = (props: AITransformationTaskpaneProps): JSX.Ele
...
+                                placeholder="Send a message."
+                                value={userInput}
+                                onChange={(e) => {
+                                    setUserInput(e.target.value);
+                                }}
+                                onKeyDown={(e) => {
+                                    if (e.key === 'Enter') {
+                                        if (!e.shiftKey) {
+                                            e.preventDefault()
+                                            submitChatInput(userInput)
```
Comment (naterush) [inline mitosheet/src/components/taskpanes/AITransformation/AITransformationTaskpane.tsx:280]: Promise

**Review by naterush — COMMENTED** [review]
Cleanup!

**Comment by aarondr77** [comment]
Random product idea (will put in better place). It would be awesome to add text competition to the chat. So if I start typing out a column header or dataframe name it will appear and I can accept it.

**Comment by aarondr77** [comment]
Using CMD + C to try and copy code in the notebook doesn't work anymore and it taking me back to Mito. 
https://user-images.githubusercontent.com/18709905/235762257-674f3a04-2469-4e8a-8025-0ea3b1060135.mov

**Comment by aarondr77** [comment]
These headers are a bit harsh. Can we try using our gray instead of black? And/or making them less bold or something
Screenshot 2023-05-02 at 3 10 42 PM

**Comment by aarondr77** [comment]
The recon is giving me incorrect information here. It says that the new dataframe has 0 rows and 0 columns

Screenshot 2023-05-02 at 3 11 57 PM

**Comment by aarondr77** [comment]
I can't find steps to replicate, but sometimes my chats are disappearing. See that the last generated code is the filter, but it's not at the bottom of the chat. It is in the step history. 
Screenshot 2023-05-02 at 3 16 46 PM
Screenshot 2023-05-02 at 3 17 11 PM

**Comment by aarondr77** [comment]
The chat history disappears when I open the AI taskpane. We should at least store the chat history between opening and closing the taskpane as long as I don't rerun the mitosheet.

**Comment by aarondr77** [comment]
Can you make sure the loading indicator is centered

Screenshot 2023-05-02 at 3 27 12 PM

**Comment by aarondr77** [comment]
We didn't specify this in the spec, but I assumed it was going to be part of this PR. Now that we are showing the chat history, I thought we were going to provide the chat history to Open AI also. See the mechanism for that here: 

https://platform.openai.com/docs/guides/chat/introduction

**Review by aarondr77 — COMMENTED** [review]
This is sweet! The biggest improvements IMO are
1) The auto execution of the code 
2) Focus handling
Pretty much, just being able to type -> see updates -> type is great.

File: mitosheet/src/components/taskpanes/AITransformation/AITransformationTaskpane.tsx [hunk 3]
```diff
@@ -255,161 +180,129 @@ const AITransformationTaskpane = (props: AITransformationTaskpaneProps): JSX.Ele
...
+                                className="ai-transformation-message ai-transformation-message-user"
+                            >
+                                <p>{taskpaneState.userInput}</p>
+                            </Row>
+                            <Row
+                                justify="start" align="center"
+                                className="ai-transformation-message ai-transformation-message-ai"
+                            >
+                                {taskpaneState.type === 'error loading completion'
+                                    ? <p>Error loading completion. This is likely because you are not connected to the internet, or there is a firewall blocking OpenAI.</p>
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/src/components/taskpanes/AITransformation/AITransformationTaskpane.tsx]: Check out that and the pro CTA.

**Comment by naterush** [comment]
There are two recon bugs you found: 
1. Not detecting series
2. Detecting the wrong df instead of the correct one

Both of these are worth fixing, but not here! We're going to do an entire product cycle on the recon, and this is a great to chance to improve it on that. I added these to a spec for the recon for ai shape, here: https://www.notion.so/trymito/Recon-for-AI-372cb0eec1e74b4083e59c9157b11f31?pvs=4

**Comment by aarondr77** [comment]
It would be nice if this text did not split in the middle of the world "folder"

Screenshot 2023-05-04 at 10 47 46 AM

**Comment by aarondr77** [comment]
This is a bit confusing. I asked it to get the most recent trip, and it generated a variable in the code, but didn't show anything in the recon or the sheet, while also saying a bunch of columns have been modified. 

Screenshot 2023-05-04 at 10 49 42 AM

**Pull request merged**